syntax_and_semantics/operators.md

```diff
@@ -21,101 +45,271 @@ struct Vector2
   def initialize(@x : Int32, @y : Int32)
   end
 
-  def +(other)
-    Vector2.new(x + other.x, y + other.y)
+  def - : self
+    Vector2.new(-x, -y)
   end
 end
 
 v1 = Vector2.new(1, 2)
-v2 = Vector2.new(3, 4)
-v1 + v2               #=> Vector2(@x=4, @y=6)
+-v1 #=> Vector2(@x=-1, @y=-2)
 ```
 
-Next follows the full list of operators with their usual meaning.
+**Note:** The operator `!` cannot be implemented as a method. It is defined by
+the compiler and can't be changed in user code.
 
-## Unary operators
+## Binary operators
 
-```crystal
-+   # positive
--   # negative
-!   # not
-~   # bitwise complement
-```
+Binary operators have two operands. Thus, a method implementation receives
+exactly one argument for the second operator. The first operand is the receiver
+`self`.
 
-These are defined without arguments. For example
+The following example demonstrates the `Vector2` type as a two-dimensional vector
+with with a binary operator method `+` for vector addition.
 
 ```crystal
 struct Vector2
-  def -
-    Vector2.new(-x, -y)
+  getter x, y
+
+  def initialize(@x : Int32, @y : Int32)
+  end
+
+  def +(other : self) : self
+    Vector2.new(x + other.x, y + other.y)
   end
 end
 
 v1 = Vector2.new(1, 2)
--v1                    #=> Vector2(@x=-1, @y=-2)
+v2 = Vector2.new(3, 4)
+v1 + v2 #=> Vector2(@x=4, @y=6)
 ```
 
-**Note:** `!` (not) cannot be defined as a method (its meaning can't be changed).
+Per convention, the return type of a binary operator should be the type of the
+first operand (the receiver): `typeof(a <op> b) == typeof(a)`
+Otherwise the assignment operator (`a <op>= b`) would unintentionally change the
+type of `a`.
+There can be reasonable exceptions though. For example in the standard library
+the float division operator `Int32#/(other : Int32)` returns `Float64`.
 
-## Binary operators
+## Ternary operators
 
-* `+` – addition
-* `-` – subtraction
-* `*` – multiplication
-* `/` – division
-* `%` – modulo
-* `&` – bitwise and
-* `|` – bitwise or
-* `^` – bitwise xor
-* `**` – exponentiation
-* `<<` – shift left, append
-* `>>` – shift right
-* `==` – equals
-* `!=` – not equals
-* `<` – less
-* `<=` – less or equal
-* `>` – greater
-* `>=` – greater or equal
-* `<=>` – comparison
-* `===` – [case equality](case.html)
-
-## Indexing
+There are no real ternary operators in Crystal.
 
-```crystal
-[]  # array index (raises on out of bounds)
-[]? # array index (nil on out of bounds)
-[]= # array index assignment
-```
+The conditional operator (`? :`) however, is technically a ternary operator.
+But it is not implemented as such, the compiler simply transforms it to an `if`
+condition.
 
-For example:
+## List of Operators
 
-```crystal
-class MyArray
-  def [](index)
-    # ...
-  end
+This list is sorted by precedence, upper entries bind stronger than lower ones.
```

Comment:
This would read more smoothly with a conjunction ("so", perhaps) after the comma.